packages/eslint-plugin/src/rules/prefer-find.ts

```diff
@@ -0,0 +1,110 @@
+import type { TSESLint } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+
+import { createRule, getParserServices } from '../util';
+
+export default createRule({
+  name: 'prefer-find',
+  meta: {
+    docs: {
+      description:
+        'Enforce the use of Array.prototype.find() over Array.prototype.filter() followed by [0] when looking for a single result',
+      requiresTypeChecking: true,
+      recommended: 'stylistic',
+    },
+    fixable: 'code',
+    messages: {
+      preferFind: 'Use .filter(...)[0] instead of .find(...)',
+      preferFindFix: 'Use .filter(...)[0] instead of .find(...)',
+    },
+    schema: [],
+    type: 'suggestion',
+    hasSuggestions: true,
+  },
+
+  defaultOptions: [],
+
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    return {
+      MemberExpression(node): void {
+        // Check if it looks like <<stuff>>[0] or <<stuff>>['0'], but not <<stuff>>?.[0]
+        if (
+          node.computed &&
+          !node.optional &&
+          node.property.type === AST_NODE_TYPES.Literal &&
```

Comment:
[Bug] We have at least one utility for this, `getStaticValue`. That way if someone does something like...

```ts
const value = 0;

values.filter(/* ... */)[value];
```

...then it'll be supported. Which is a good test case to add.

Here and the other `type === AST_NODE_TYPES.Literal`.

Reply:
Changed.... though, I also wonder whether we should try to account for this case (which doesn't presently work)
```ts
declare const filterLiteral: 'filter';
[1, 2, 3][filterLiteral](callback)[0];
```
This would require the type checker, of course. Not sure if that's a significant tradeoff or anything.

Reply:
nah we aren't looking to support weird shit like that. Variables as array indexers are normal - variables as indexers for method names are not.